### ISSUE TITLE
Feature Request : Support JSX

### ISSUE THREAD

**Issue by dylans** [dojo/widget-core#469]
@sebilasse commented on [Fri Apr 28 2017](https://github.com/dojo/cli/issues/144)

Wondered how my "html preferring" users could write dojo2 widgets in a declarative way and just read https://medium.com/maquette-news/maquette-2-2-now-supports-jsx-7da7b0a1dc98
This is so nice. I could think of a great benefit for dojo users.

---

@dylans commented on [Fri Apr 28 2017](https://github.com/dojo/cli/issues/144#issuecomment-298147394)

We have support for this in a branch on one of our contributor's repos. While we think jsx is a step in the wrong direction compared to hyperscript with regards to functional reactive programming, and we're trying to avoid the temptation of supporting too many different ways of doing things (which can lead to more complexity), we will reconsider adding this as option before the release of Dojo 2.

---

@dylans commented on [Sat Apr 29 2017](https://github.com/dojo/cli/issues/144#issuecomment-298166412)

I'm going to move this issue to widget-core since that is where it would be supported

**Comment by kitsonk** [comment]
There is good reason why end-users should be abstracted from the DOM.  I think that JSX is a significant step backwards in usability and maintainability of code.  We are still finding use cases where we need to continue to abstract end developers from the DOM (and therefore HTML) and JSX would make it very difficult to assist developers in following those patterns to make well behaved widgets.  HTML, in my opinion, is a _horrible_ way to express a widget (or a DOM node).  It is only the familiarity that deludes us into believing it is a better way.

My strong opinion would be that this is a great opportunity for the wider community to consider if there is a good way to create well behaved widgets that fit into widget core with JSX or TSX, but that for the team working on `widget-core` that there are other items that need to be addressed before the release of a higher priority.

**Comment by kitsonk** [comment]
@matt-gadd has done a POC of supporting TSX in `widget-core`.  The raw functionality appears to be fairly straight forward and appears to integrate well into the concepts in `widget-core`.

I still have reservations about what _I_ would choose as a developer of maintainable widgets, preferring a functional mechanism instead of an HTML _like_ language, as I still feel it is easier to understand that intent of the code.

We clearly need to do further investigations to see if we can leverage the full benefits of the widgeting system in applications that takes advantage of the APIs and TypeScript in a way that generates well behaved widgets.  This appears that it is likely the case, but we should validate that and understand any limitations/restrictions/oddities before committing to deliver support for TSX as an "official" way of authoring widgets.

**Comment by sebilasse** [comment]
@kitsonk 
Agreed now, especially to the cursive _I_ aka `user first` - 
many people might have huge libraries of "html and css" components which are well behaved at least for semantics and style). They might want to use them that as a starter point.

For political reasons I would never use TSX because I never use Facebook stuff (or Walmart [hapi] or so) but I would not expect such a clear opinion from my users ;)
The question is weather TSX (which I don't like personally as well) would be the best fitting declarative syntax.
Personally I like the idea of http://riotjs.com ...
What I wrote as my own template system last weeks is the Django (Swig/Twig etc.) syntax as an ES6 Template system (including fast express.js rendering from a function - this function is precompiled on first rendering and cached, there is no need for "building"), I'd try a POC for a riot-like syntax now...

**Comment by kitsonk** [comment]
Well _JSX_ might be facebook, but _TSX_ is TypeScript.  It would be difficult, if not impossible for us to provide a good user experience without using _TSX_.

The only other option that could be considered is [tagged template literals](https://developer.mozilla.org/en/docs/Web/JavaScript/Reference/Template_literals) which would desugar to `WNode`s and `HNode`s... something like:

```ts
class MyWidget extends WidgetBase<WidgetProperties> {
    render() {
        return vTag`<div bind=${this} class=${this.classes(css.main)}>${this._text}</div>`;
    }
}
```

By that point you are pretty much back at TSX.

**Comment by agubler** [comment]
@sebilasse a pull request is up to add support for JSX https://github.com/dojo/widget-core/pull/475

**Comment by sebilasse** [comment]
@kitsonk @agubler supercool.
@kitsonk :

> The only other option that could be considered is tagged template literals

Exactly,
FYI : This is also what I did with "the Django (Swig/Twig etc.) syntax as an ES6 Template system" ...
A snippet for the IndieAuth layout
```html
<body class="indieauth ${baseClass}">
  <!-- Scroll and Mobile Navs -->
  include`./_partial/nav.html`
  <!-- Page Contents -->
  <div class="${color||'black'} pusher">
    <header class="ui ${color||'black'} inverted vertical masthead center aligned segment">
      include`./_partial/header.html`
    </header>
    <main>
      block`content`
        <!-- THIS WILL BE REPLACED WITH THE CONTENT BY THE FILE WICH IS RENDERED AND "EXTENDS" THIS FILE -->
      endblock`content`
    </main>
    <footer class="ui ${color||'black'} inverted vertical footer segment">
      include`./_partial/footer.html`
    </footer>
  </div>
  <script src="library/main.js"></script>
  list(scripts)`<script src="${value}" type="text/javascript"></script>`
```

Everything will compile to a function which could be precompiled (that happens automatically at first node.js render) - a larger page with much logic takes then e.g. about 60ms for first render and 12ms when precompiled and cached.
Basically I try now to replace the `include` parts by webcomponents.
As said it is important to me that they are 'curlable' as well and so I render the app-root with node.js/jsdom as Kitson explained. 

There will be much components which do not need any logic (just CSS / https://semantic-ui.com) and so I'd prefer to write HTML.
On the other side there will be components which should have well behaving logic and so I'd write them as dojo widgets. In the sense of "progressive enhancement" another approach for the JS layer could be the theory behind https://github.com/reactjs/react-magic ("diff" the old HTML with the new HTML seems very attracting to me).
